File: 1023-time-based-key-value-store/time-based-key-value-store.java
class TimeMap {

    Map<String , TreeMap<Integer,String>> tmap;
    public TimeMap() {
        tmap= new HashMap<String,TreeMap<Integer,String>>();
    }
    
    public void set(String key, String value, int timestamp) {
        if(!tmap.containsKey(key)){
            tmap.put(key, new TreeMap<Integer,String>());
        }
        tmap.get(key).put(timestamp,value);
    }
    
    public String get(String key, int timestamp) {
        if(!tmap.containsKey(key)){
            return "";
        }

        Integer floorval = tmap.get(key).floorKey(timestamp);

        if(floorval !=null){
            return tmap.get(key).get(floorval);
        }

        return "";
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */

File: 1023-time-based-key-value-store/time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        self.time_map = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.time_map:
            self.time_map[key] = SortedDict()
        self.time_map[key][timestamp] = value

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.time_map:
            return ""
        ind = self.time_map[key].bisect_right(timestamp)
        if ind == 0:
            return ""
        return self.time_map[key].peekitem(ind - 1)[1]


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
